### ISSUE TITLE
Save a copy of all source files

### ISSUE THREAD

**Issue by atasak** [dsherret/ts-simple-ast#182]
Source files can be copied and saved per source file. Generated code can be emitted to a directory as .d.ts and .js files. It would be neat to be able to emit generated code to another directory as typescript code. 

Use case example: 
- Read code from `src/genSrc/`
- Mainpulate code
- Emit code as plain typescript in `src/genDist/`
- Emit declarations of generated code in `src/genSrc` so the code _before the manipulation_ may use the to be generated methods already
- Use the code in `src/genDist/` in the rest of the code in `src/*/`
- Possibly add a file watcher to automatically update the generated code when source code is altered

Maybe it is possible to refactor the emit method a bit? e.g.:

```typescript
ast.emit({
    'emitCompiledJavascript': 'dist/genDist',
    'emitDeclaration': 'src/genSrc',
    'emitTypescript': 'src/genDist',
});
```

**Comment by dsherret** [comment]
I agree, this would be useful! It seems like this would work well with the new Directories feature. I've opened #184 and #183 which should help do this.

I think the final code would end up something like:

```ts
// does not work yet

// assuming "src" is the root
const genSrc = ast.getDirectoryOrThrow("genSrc");
const genDist = genSrc.copy("../genDist");

// do manipulations on genDist's source files here

genDist.emit({
    // would use the defaults from the compiler options, but you can override them here like so...
    declarationDir: "../genSrc",
    outDir: "../../dist/genDist"
});
```

**Comment by atasak** [comment]
That would work, assuming you could also save a directory. In the example above, you do not need the copy, as the emit function already emits to a different out folder. What I would be interested in is:

```
// does not work yet

// assuming "src" is the root
const genSrc = ast.getDirectoryOrThrow("genSrc");
const genDist = genSrc.copy("../genDist");

// do manipulations on genDist's source files here

// Add the ability to do this:
genDist.saveAll();

genDist.emit({
    // would use the defaults from the compiler options, but you can override them here like so...
    declarationDir: "../genSrc",
    outDir: "../../dist/genDist"
});
```

**Comment by dsherret** [comment]
Sure, opened #185.

**Comment by dsherret** [comment]
This is implemented in 4.2.1. You should be able to do this:

```typescript
const genSrc = ast.getDirectoryOrThrow("genSrc");
const genDist = genSrc.copy("../genDist");

// do manipulations on genDist's source files here

await genDist.saveUnsavedSourceFiles();

const result = await genDist.emit({
    declarationDir: "../genSrc",
    outDir: "../../dist/genDist"
});
// if you need the emitted file paths, they exist in result.getOutputFilePaths()
```

Let me know if this solves the issue or if you have any questions. Thanks!

**Comment by atasak** [comment]
Thanks!
